# ISAE: Identify super accessibility elements  

## 1.Introduction  
The software was used to identify super accessibility elements (SAEs) using ATAC-seq data, which were open chromatin regions in promoters or enhancers that are strongly bound simultaneously by multiple lineage-specific TFs. In addition to this, the software can also associate SAEs with gene expression and identify differential SAEs between different tissues/cell stages and types considering the change both in ATAC signal and TF footprint binding.   

## 2.Dependencies  
python3  
R > v3.4  
perl > v5  
bedtools > v2  
samtools > v1  
macs2 > v2  
pyDNase v0.030  
HOMER > v4  
wigToBigWig  
bigWigToBedGraph  

Most of these dependencies can be installed at once using conda: http://conda.pydata.org/miniconda.html  
## 3.Installation  
```
git clone https://github.com/MingyangHu/ISAE.git
export PATH=$PATH:/path/to/SAE/bin
```

## 4.Usage  
### 1>Idnetify_SAE.py  
The script was used to call ATAC broad peak, identify footprints in ATAC peak and super accessibility elements.    
```
Idnetify_SAE.py [options] -t Input_align_file_for_call_peak -b Bam_file -n Experiment_name  -o Output_directory  -g Effective_genome_size --g_file the_chrom_size_file  
```
* Explanation of parameters    
-n                          Experiment name, which will be used to generate output file names.    
-p                          Number of processes to use.    
--outdir                    Path to output directory.    
--Broad_peak                If --macs2_call_peak choose "off",please provide a broad peak bed file.    
--footprint_file            If choose "off" for --wellington_footprints, please provide the footprint bed file with score information.  
###############Macs2 call peak  
-t                          The algin file of ATAC-seq data. It is strongly recommended to use the tn5.tagAlign file generated by the ENCODE ATAC-seq pipeline (https://github.com/kundajelab/atac_dnase_pipelines), in which all reads aligning to the + strand were offset by +4 bp, and all reads aligning to the – strand were offset −5 bp.  
-f                          Format of algin file,{AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE}. The default AUTO option will let MACS decide which format. Default:AUTO     
-g                          Effective genome size.It can be 1.0e+9 or 1000000000, or shortcuts:'hs' for human (2.7e9), 'mm' for mouse(1.87e9), 'ce' for C. elegans (9e7) and 'dm' for fruitfly (1.2e8), Default:hs.    
--pvalue                    Pvalue threshold to filter peak, Default:5 (means P value < 1E-5).    
--macs2_call_peak           Whether using macs2 to call peak,can choose "on" or "off". Default:on    
###############Footprint analysis  
-b                          Bam file of ATAC-seq data.   
-A                          ATAC-seq mode (default: True).      
--wellington_footprints     Whether to execute footprint analysis,can choose "on" or "off".Default:on  
--Dnase_wig_To_bw           If choose "on" for --wellington_footprints, this parameter can choose "on" or "off" to decide whether to creat bw file of cut information from a ATAC-seq experiment. Default:on  
###############Identify SAE  
--c_proximal                the cut off of footrpitn plvalue and footprint number to identify super proximal ATAC peak  
--c_distal                  the cut off of footrpitn plvalue and footprint number to identify super distal ATAC peak  
--TSS                       the promoter file of gene  
--houseKeeping              the promoter file of houseKeeping gene  
--proximal                  whether to identify proximal super peak,can choose "on" or "off".Default:on  
--distal                    whether to identify distal super peak,can choose "on" or "off".Default:on  
-h                          Print this help menu  
* Usage examples  
Ⅰ.The following is an example using align file as the input, the analysis process includes peak calling, footprint analysis and SAE identification.  
 ```  
Idnetify_SAE.py -t C2C12-DM60h-ATAC.nodup.tn5.tagAlign.gz -g mm -n C2C12_DM60h_ATAC -A -p 8 -b C2C12-DM60h-ATAC.nodup.bam --outdir /path/to/C2C12_DM60h  --g_file mm11.chrom.sizes --Pvalue 10 --c_distal 10 --c_proximal 10   --TSS C2C12_DM60h_promoter.bed  --houseKeeping mouse_houseKeeping_gene_TSS_region.bed    
 ```  
### 2>link_ATAC_peak_with_gene.py  
The script was used to assign super and typical accessibility elements to each expressed gene and compare the gene expression between the two group genes.    
```
link_ATAC_peak_with_gene.py -s super peak bed file -t typical peak bed file -g  gene tss bed file  -G gff annotation file  -k houseKeepgene_ID_list -h Print this help menu  
```
### 3>DifferentialSAE.py  
The script was used to identify differential SAEs between different tissues/cell stages and types considering the change both in ATAC signal and TF footprint binding.     
```
DifferentialSAE.py  -pt  Treatment_ATAC_broad_peak_bed_file -pc   Control_ATAC_broad_bed_file -bt  Treatment_bam_file  -bc  Control_bam_file  -st  Treatment_SAE_bed_file  -sc  Control_SAE_bed_file -nt Treatment_sample_name -nc Control_sample_name -o output_path  
```
* Explanation of parameters  
-pt ATAC broad peak bed file of treatment group  
-pc ATAC broad peak bed file of control group  
-bt ATAC bam file of treatment group  
-bc ATAC bam file of control group  
-st SAE  bed file of treatment group -st SAE  bed file of control group  
-nt sample name of treatment group  
-nc sample name of control group  
-o Path to output folder  
-h Print this help menu  
